cpp/src/arrow/util/small_vector.h

```diff
@@ -0,0 +1,487 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <cassert>
+#include <cstddef>
+#include <initializer_list>
+#include <iterator>
+#include <limits>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "arrow/util/launder.h"
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace internal {
+
+template <typename T>
+struct StaticVectorMixin {
+  // properly aligned uninitialized storage for N T's
+  using storage_type = typename std::aligned_storage<sizeof(T), alignof(T)>::type;
+
+  static T* ptr_at(storage_type* p, size_t i) {
+    return launder(reinterpret_cast<T*>(&p[i]));
+  }
+
+  static constexpr const T* ptr_at(const storage_type* p, size_t i) {
+    return launder(reinterpret_cast<const T*>(&p[i]));
+  }
+
+  static void move_storage(storage_type* src, storage_type* dest, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      T* src_item = ptr_at(src, i);
+      T* dest_item = ptr_at(dest, i);
+      new (dest_item) T(std::move(*src_item));
+      src_item->~T();
+    }
+  }
+
+  static void destroy_storage(storage_type* p, size_t n) {
+    for (size_t i = 0; i < n; ++i) {
+      ptr_at(p, i)->~T();
+    }
+  }
+};
+
+template <typename T, size_t N, bool NonTrivialDestructor>
+struct StaticVectorStorageBase : public StaticVectorMixin<T> {
+  using typename StaticVectorMixin<T>::storage_type;
+
+  storage_type static_data_[N];
+  size_t size_ = 0;
+
+  void destroy() {}
+};
+
+template <typename T, size_t N>
+struct StaticVectorStorageBase<T, N, true> : public StaticVectorMixin<T> {
+  using typename StaticVectorMixin<T>::storage_type;
+
+  storage_type static_data_[N];
+  size_t size_ = 0;
+
+  ~StaticVectorStorageBase() noexcept { destroy(); }
+
+  void destroy() noexcept { this->destroy_storage(static_data_, size_); }
+};
+
+template <typename T, size_t N, bool D = !std::is_trivially_destructible<T>::value>
+struct StaticVectorStorage : public StaticVectorStorageBase<T, N, D> {
+  using Base = StaticVectorStorageBase<T, N, D>;
+  using typename Base::storage_type;
+
+  using Base::size_;
+  using Base::static_data_;
+
+  StaticVectorStorage() noexcept = default;
+
+  T* data_ptr() { return this->ptr_at(static_data_, 0); }
+
+  constexpr const T* const_data_ptr() const { return this->ptr_at(static_data_, 0); }
+
+  void bump_size(size_t addend) {
+    assert(size_ + addend <= N);
+    size_ += addend;
+  }
+
+  void ensure_capacity(size_t min_capacity) { assert(min_capacity <= N); }
+
+  void reduce_size(size_t reduce_by) {
+    assert(reduce_by <= size_);
+    size_ -= reduce_by;
+  }
+
+  void move_from(StaticVectorStorage&& other) noexcept {
+    size_ = other.size_;
+    this->move_storage(other.static_data_, static_data_, size_);
```

Comment:
IIUC this doesn't call destructors for any objects currently in `this->static_data_`

Reply:
Indeed you need to destroy the current storage explicitly if needed (by calling `destroy()`).